pkg/cmd/cli/cmd/importimage.go

```diff
@@ -67,45 +68,147 @@ func RunImportImage(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, arg
+	if len(ref.ID) > 0 {
+		return fmt.Errorf("if you want to import an image by ID, use --from=%s %s", from, ref.AsRepository().Exact())
+	}
+	// apply the default tag
+	if !all && len(ref.Tag) == 0 {
+		ref.Tag = imageapi.DefaultImageTag
+	}
+	name := ref.Name
+
 	imageStreamClient := osClient.ImageStreams(namespace)
-	stream, err := imageStreamClient.Get(streamName)
+	stream, err := imageStreamClient.Get(name)
 	if err != nil {
-		if len(from) == 0 || !errors.IsNotFound(err) {
+		if !errors.IsNotFound(err) {
 			return err
 		}
-		if !confirm {
-			return fmt.Errorf("the image stream does not exist, pass --confirm to create")
+		if len(from) == 0 && !confirm {
+			return fmt.Errorf("no image stream named %q exists, pass --confirm to create and import", name)
 		}
-		stream = &imageapi.ImageStream{
-			ObjectMeta: kapi.ObjectMeta{Name: streamName},
-			Spec:       imageapi.ImageStreamSpec{DockerImageRepository: from},
+
+		if len(ref.Tag) == 0 && all {
+			stream = &imageapi.ImageStream{
+				ObjectMeta: kapi.ObjectMeta{Name: name},
+				Spec:       imageapi.ImageStreamSpec{DockerImageRepository: from},
+			}
+		} else {
+			stream = &imageapi.ImageStream{
+				ObjectMeta: kapi.ObjectMeta{Name: name},
+				Spec: imageapi.ImageStreamSpec{
+					Tags: map[string]imageapi.TagReference{
+						ref.Tag: {
+							From: &kapi.ObjectReference{
+								Kind: "DockerImage",
+								Name: from,
+							},
+						},
+					},
+				},
+			}
 		}
+
 	} else {
 		if len(stream.Spec.DockerImageRepository) == 0 && len(stream.Spec.Tags) == 0 {
 			return fmt.Errorf("image stream has not defined anything to import")
 		}
-		if len(from) != 0 {
-			if from != stream.Spec.DockerImageRepository {
+		if len(ref.Tag) == 0 && all {
+			if len(from) != 0 && from != stream.Spec.DockerImageRepository {
 				if !confirm {
+					if len(stream.Spec.DockerImageRepository) == 0 {
+						return fmt.Errorf("the image stream does not currently import an entire Docker repository, pass --confirm to update")
+					}
 					return fmt.Errorf("the image stream has a different import spec %q, pass --confirm to update", stream.Spec.DockerImageRepository)
 				}
 				stream.Spec.DockerImageRepository = from
 			}
+		} else {
+			var tag imageapi.TagReference
+			if existing, ok := stream.Spec.Tags[ref.Tag]; ok {
+				tag = existing
+				delete(tag.Annotations, imageapi.DockerImageRepositoryCheckAnnotation)
+			} else {
+				tag = imageapi.TagReference{
+					From: &kapi.ObjectReference{
+						Kind: "DockerImage",
+						Name: from,
+					},
+				}
+			}
+			stream.Spec.Tags[ref.Tag] = tag
 		}
 	}
 
-	if stream.Annotations != nil {
-		delete(stream.Annotations, imageapi.DockerImageRepositoryCheckAnnotation)
+	// Attempt the new, direct import path
+	isi := &imageapi.ImageStreamImport{
+		ObjectMeta: kapi.ObjectMeta{
+			Name:            stream.Name,
+			Namespace:       namespace,
+			ResourceVersion: stream.ResourceVersion,
+		},
+		Spec: imageapi.ImageStreamImportSpec{Import: true},
```

Comment:
Changing roles is surprising, hence it's confirm.  Refreshing an existing tag is not surprising, therefore it doesn't require confirm.  Saying "import" but then creating a new image stream is surprising, therefore it requires confirm.